File: week-4/implement-stack-using-queues/implement-stack-using-queues.js
/**
 * Initialize your data structure here.
 */
const Queue = function () {
    this.front = 0;
    this.queue = [];
}

Queue.prototype.push_to_back = function(val) {
    this.queue.push(val);
}

Queue.prototype.pop = function() {
    const val = this.queue[this.front];
    ++this.front;
    return val;
}

Queue.prototype.peek = function() {
    return this.queue[this.front];
}

Queue.prototype.is_empty = function() {
    return this.queue.slice(this.front).length === 0;
}

Queue.prototype.size = function() {
    return this.queue.slice(this.front).length;
}

var MyStack = function() {
    this.q = new Queue();
};

/**
 * Push element x onto stack. 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function(x) {
    this.q.push_to_back(x);
    let size = this.q.size();
    while (size > 1) {
        this.q.push_to_back(this.q.pop());
        --size;
    }
};

/**
 * Removes the element on top of the stack and returns that element.
 * @return {number}
 */
MyStack.prototype.pop = function() {
    return this.q.pop();
};

/**
 * Get the top element.
 * @return {number}
 */
MyStack.prototype.top = function() {
    return this.q.peek();
};

/**
 * Returns whether the stack is empty.
 * @return {boolean}
 */
MyStack.prototype.empty = function() {
    return this.q.is_empty();
};

/** 
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: week-4/implement-stack-using-queues/implement-stack-using-queues.py
class Queue:
    def __init__(self):
        self.queue = []
        self.front = 0
        
    def push_to_back(self, val):
        self.queue.append(val)
        
    def size(self):
        return len(self.queue[self.front:])
    
    def pop(self):
        val = self.queue[self.front]
        self.front += 1
        return val
    
    def peek(self):
        return self.queue[self.front]
        
    def is_empty(self):
        return len(self.queue[self.front:]) == 0

class MyStack:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.q = Queue()

    def push(self, x: int) -> None:
        """
        Push element x onto stack.
        """
        self.q.push_to_back(x)
        size = self.q.size()
        while size > 1:
            self.q.push_to_back(self.q.pop())
            size -= 1

    def pop(self) -> int:
        """
        Removes the element on top of the stack and returns that element.
        """
        return self.q.pop()
        

    def top(self) -> int:
        """
        Get the top element.
        """
        return self.q.peek()

    def empty(self) -> bool:
        """
        Returns whether the stack is empty.
        """
        return self.q.is_empty()


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()
